README.md
# Transport Challenge

In the Transport Challenge API, the [**Magnebot**](https://github.com/alters-mit/magnebot) must transport **target objects** (small objects scattered throughout the scene) with the aid of **containers** (box-shaped objects without lids that can hold target objects) to the **goal zone** (a circle defined by a position and a radius in the center of a room and the scene).

If all of the target objects are in the goal zone, the task is successful.

**[Read the API documentation here.](https://github.com/alters-mit/transport_challenge/blob/main/doc/transport_controller.md)**

The Transport Challenge API is an extension of the [Magnebot API](https://github.com/alters-mit/magnebot) which in turn is built on the [TDW simulation platform](https://github.com/threedworld-mit/tdw).

<img src="doc/images/api_hierarchy.png" style="zoom:50%;" />

# Installation

Check if you've already installed `magnebot` or `tdw` on your computer: if `pip3 show magnebot` or `pip3 show tdw` prints any text, then the module is installed.

If `magnebot` or `tdw` is *not* already installed on this computer:

1. [Download TDW build v1.8.7](https://github.com/threedworld-mit/tdw/releases/tag/v1.8.7)
2. `git clone https://github.com/alters-mit/transport_challenge.git`
3. `cd transport_challenge` 
4. `pip3 install -e .` This will install all underlying modules, including `magnebot` and `tdw`
5. Make sure that the underlying TDW system is working correctly. See: [Getting Started With TDW](https://github.com/threedworld-mit/tdw/blob/master/Documentation/getting_started.md#Requirements) ***Ignore warnings about upgrading TDW!***

If `magnebot` or `tdw` *is* already installed on this computer:

1. `pip3 uninstall magnebot`
2. `pip3 uninstall tdw`
3. `pip3 unintall ikpy`
4. [Download TDW build v1.8.7](https://github.com/threedworld-mit/tdw/releases/tag/v1.8.7)
5. `git clone https://github.com/alters-mit/transport_challenge.git`
6. `cd transport_challenge` 
7. `pip3 install -e .` This will install all underlying modules, including `magnebot` and `tdw`
8. Make sure that the underlying TDW system is working correctly. See: [Getting Started With TDW](https://github.com/threedworld-mit/tdw/blob/master/Documentation/getting_started.md#Requirements) ***Ignore warnings about upgrading TDW!***

# Usage

1. Run this controller: `python3 controller.py` If you're running on an IBM server, you'll need to set a custom S3 bucket for the scenes, models, and materials: `TRANSPORT_CHALLENGE=https://bucket_url python3 controller`. If this is a local path, set the variable to the *absolute* path, i.e. `TRANSPORT_CHALLENGE=file:///home/<user>/model_library python3 controller`

```python
from transport_challenge import Transport

m = Transport()
# Initializes the scene.
status = m.init_scene(scene="2a", layout=1)
print(status)  # ActionStatus.success

# Prints a list of all container IDs.
print(m.containers)
# Prints a list of all target object IDs.
print(m.target_objects)

m.end()
```

2. [Launch the TDW build.](https://github.com/threedworld-mit/tdw/blob/master/Documentation/getting_started.md)

# Documentation

- **[API documentation](https://github.com/alters-mit/transport_challenge/blob/main/doc/transport_controller.md)**
- [Changelog](https://github.com/alters-mit/transport_challenge/blob/main/doc/changelog.md)

# Example controllers

- [This controller](https://github.com/alters-mit/transport_challenge/tree/main/controllers/examples/single_room.py) is an *example use-case*. It uses very naive logic to navigate (it assumes that everything is in the same room and that there aren't obstructions between objects) but it should be a good example of how to use this API.
- [This controller](https://github.com/alters-mit/transport_challenge/tree/main/controllers/demos/demo.py) is a *promo controller*. It is visually indicative of an actual use-case and includes an overhead camera so that it's easy to see what's going on. However, this controller includes a lot of code that you shouldn't add to your controller because it's unnecessary, inflexible, and slow.
- [These controllers](https://github.com/alters-mit/transport_challenge/tree/main/controllers/tests) are *test controllers*. They are meant only for testing the API.


doc/changelog.md
# Changelog

## 0.3.7

- Fixed: Crash because Magnebot 1.1.1 is installed

## 0.3.6

- Requires: Magnebot 1.1.1 (was 1.1.2, which was incorrect)
- Requires: ikpy 3.1
- Added install instructions to the README for downgrading `magnebot` and `tdw`

## 0.3.5

- Fixed: README points to the wrong version of the build.

## 0.3.4

- Added: `util/target_object_sizes.py` Print the size of each target object.
- Documentation now includes the capacity of the containers.

## 0.3.3

- Requires: TDW 1.8.7.0

## 0.3.2

- **The Transport Challenge API requires Magnebot v1.1.2 and should NOT be upgraded past this version!**
  - pip will install the correct version of Magnebot
  - Added: `Transport.MAGNEBOT_VERSION` The required version of the Magnebot Python module
- Added: `promo_controller.y` Base class for promo controllers
- Added: `failure.py` Demo of the Magnebot failing a task
- Added: `success.py` Demo of the Magnebot succeeding at a task

## 0.3.1

- Fixed: Object URLs are not set correctly if the `TRANSPORT_CHALLENGE` environment variable is provided
- Fixed: `promo.py` doesn't work as intended due to collisions with containers

## 0.3.0

- Requires: Magnebot 1.1.0 (see changelog notes regarding collision detection)
  - Added optional parameter `stop_on_collision` to `move_by()`, `move_to()`, `turn_by()` and `turn_to()` Set this to False to ignore collision detection during the action
- Fixed: Crash-to-desktop because a container gets caught in the Magnebot's wheels. Now, the Magnebot will stop moving as soon as a wheel collides with a container.
- Added: `tests/container_collision.py`

## 0.2.3

- Added optional parameter `check_pypi_version` to the constructor
- Simplified the format of this changelog

## 0.2.2

- Added: Set the `TRANSPORT_CHALLENGE` environment variable to use an alternative S3 bucket URL.

## 0.2.1

- Added optional parameter `random_seed` to `init_scene()` to reset the random seed.

## 0.2.0

- Added `scipy` as a required module
- There are always 8 target objects in the scene (previously, there are 8-12 target objects)
- Target objects can spawn in any room (previously, they spawned in the same room)
- Fixed: Target objects sometimes spawn on the edges of the occupancy map
- Fixed: Containers sometimes spawn in very small rooms
- Added optional parameter `fov` to the constructor
- Simplified the logic of `single_room.py` and made it work with the changes in this update
- `promo.py` uses init data from `init.json` to reload its state

## 0.1.7

- Fixed: Default value of `img_is_png` is True (it's now False)

## 0.1.6

- Fixed: JSONDecodeError because version check doesn't work (version check for this module has been removed)

## 0.1.5

- Added license
- Default value of `skip_frames` is 10 (was 20)
- Renamed `controllers/demo/demo.py` to `controllers/promos/promo.py`

## 0.1.4

- `put_in()` will immediately stop moving the arm holding the object if the magnet or the object is within the container.
- Improved the speed of `put_in()` and `pour_out()`.
- Fixed: `put_in()` is too slow because the arm holding the object isn't sufficiently above the container.

## 0.1.3

- Added optional parameter `skip_frames` (required in Magnebot 0.4.0)
- Improved the speed of `put_in()`.
- Fixed: `pick_up()` doesn't reset the arm if the magnet fails to grasp the target object.
- Fixed: `pick_up()` often misses the object (it now aims for positions on the top of the object).
- Fixed: `put_in()` sometimes doesn't stop the Magnebot's wheels from turning.
- Fixed: `put_in()` often aims for the wrong target position above the container.
- Fixed: `pour_out()` sometimes doesn't stop the Magnebot's wheels from turning.
- Fixed: while objects are being dropped into a container during `put_in()`, they sometimes glitch (now they use the `discrete` collision detection mode).
- Fixed: `single_room.py` doesn't work. Added simple navigation to make multiple attempts to pick up the target object.
- Removed: `level.py`

## 0.1.2

- Set `launch_build` default value to False (was True)
- Fixed: `put_in()` often thinks that an object isn't in a container when it actually is.
- Fixed: `put_in()` sometimes fails because the magnet holding the target object is in the way. Now, the wrist of that magnet will tilt down, allowing for a clearer trajectory.
- Fixed: Sometimes in `put_in()` the arm that was holding the target object gets caught on the container. Now, the arm that was holding the target object resets first, rather than vice versa.
- Backend:
  - `_get_objects_in_container()` uses trigger event data (`self._trigger_events`) to determine if an object is in a container rather than `Overlap` output data. It no longer advances the simulation an extra frame.

## 0.1.1

- Added optional parameter `img_is_png` to the constructor
- Renamed `num_actions` to `action_cost`
- Added: `get_target_objects_in_goal_zone()`
- Cleaned up the code of `single_room.py`
- Added links to controllers in README

controllers/tests/fill_and_pour.py
import numpy as np
from tdw.tdw_utils import TDWUtils
from magnebot import Arm, ActionStatus
from transport_challenge import Transport


class FillAndPour(Transport):
    """
    Test whether the Magnebot can put objects in a container, move them around the scene, pour out the container.
    """

    def init_scene(self, scene: str = None, layout: int = None, room: int = None,
                   goal_room: int = None, random_seed: int = 0) -> ActionStatus:
        origin = np.array([0, 0, 0])
        commands = [{"$type": "load_scene",
                     "scene_name": "ProcGenScene"},
                    TDWUtils.create_empty_room(12, 12)]
        self._add_container(model_name="basket_18inx18inx12iin",
                            position={"x": 0.354, "y": 0, "z": 0.549},
                            rotation={"x": 0, "y": -70, "z": 0})
        # Add target objects in a circle.
        num_objects = 10
        d_theta = 360 / num_objects
        theta = d_theta / 2
        pos = np.array([2, 0, 0])
        for j in range(num_objects):
            object_position = TDWUtils.rotate_position_around(origin=origin, position=pos, angle=theta)
            self._add_target_object("jug05",
                                    position=TDWUtils.array_to_vector3(object_position))
            theta += d_theta
        commands.extend(self._get_scene_init_commands())
        resp = self.communicate(commands)
        self._cache_static_data(resp=resp)
        # Wait for the Magnebot to reset to its neutral position.
        status = self._do_arm_motion()
        self._end_action()
        return status


if __name__ == "__main__":
    m = FillAndPour(launch_build=False, random_seed=0)
    m.init_scene()
    # Pick up the container.
    m.pick_up(target=m.containers[0], arm=Arm.right)

    i = 0
    for object_id in m.target_objects:
        # Every n objects, go to the center of the room and pour out the container.
        if i > 0 and i % 4 == 0:
            m.move_to({"x": 0, "y": 0, "z": 0})
            s = m.pour_out()
            assert s == ActionStatus.success, s
        m.move_to(target=object_id)
        # Pick up the target object.
        m.pick_up(target=object_id, arm=Arm.left)
        # Put the object in the container.
        s = m.put_in()
        assert s == ActionStatus.success, s
        i += 1
    m.move_to({"x": 0, "y": 0, "z": 0})
    s = m.pour_out()
    assert s == ActionStatus.success, s
    m.move_by(-1)
    m.end()


controllers/tests/container_collision.py
from tdw.tdw_utils import TDWUtils
from magnebot import ActionStatus
from transport_challenge import Transport


class ContainerCollision(Transport):
    """
    Test collisions between the Magnebot and a container.
    """

    def init_scene(self, scene: str = None, layout: int = None, room: int = None,
                   goal_room: int = None, random_seed: int = 0) -> ActionStatus:
        commands = [{"$type": "load_scene",
                     "scene_name": "ProcGenScene"},
                    TDWUtils.create_empty_room(12, 12)]
        self._add_container(model_name="basket_18inx18inx12iin",
                            position={"x": 0.235, "y": 0, "z": 1.5})
        self._add_container(model_name="basket_18inx18inx12iin",
                            position={"x": 0.235, "y": 1, "z": 1.25})
        commands.extend(self._get_scene_init_commands())
        resp = self.communicate(commands)
        self._cache_static_data(resp=resp)
        # Wait for the Magnebot to reset to its neutral position.
        status = self._do_arm_motion()
        self._end_action()
        return status


if __name__ == "__main__":
    m = ContainerCollision(launch_build=False, random_seed=0, skip_frames=0)
    m.init_scene()
    print(m.move_by(8))
    print(m.turn_by(70))
    print(m.turn_by(-70))
    print(m.move_by(-8))
    m.end()


util/target_object_sizes.py
from csv import DictReader
import numpy as np
from tdw.librarian import ModelLibrarian
from tdw.tdw_utils import TDWUtils
from transport_challenge.paths import TARGET_OBJECTS_PATH

"""
Prints the sizes of each of the target objects.
"""

lib = ModelLibrarian()
with TARGET_OBJECTS_PATH.open("rt", encoding="utf-8") as f:
    reader = DictReader(f)
    for row in reader:
        record = lib.get_record(row["name"])
        # Get the bounds dimensions.
        width = np.linalg.norm(TDWUtils.vector3_to_array(record.bounds["left"]) -
                               TDWUtils.vector3_to_array(record.bounds["right"]))
        height = np.linalg.norm(TDWUtils.vector3_to_array(record.bounds["top"]) -
                                TDWUtils.vector3_to_array(record.bounds["bottom"]))
        length = np.linalg.norm(TDWUtils.vector3_to_array(record.bounds["front"]) -
                                TDWUtils.vector3_to_array(record.bounds["back"]))
        # Get the size times the scale.
        size = np.array([width, height, length]) * float(row["scale"])
        print(row["name"], size)


controllers/promos/failure.py
from magnebot import ActionStatus, Arm
from transport_challenge.promo_controller import PromoController


class Failure(PromoController):
    """
    Demo of a Magnebot trying an failing to do a task.
    """

    def init_scene(self, scene: str = "1a", layout: int = 1, room: int = 0, goal_room: int = None,
                   random_seed: int = 0) -> ActionStatus:
        super().init_scene(scene=scene, layout=layout, room=room, random_seed=random_seed)
        # Move a target object under a chair and another target object nearby on the floor.
        self._next_frame_commands.extend([{"$type": "teleport_object",
                                           "id": self.target_objects[0],
                                           "position": {"x": -0.689, "y": 0, "z": 2.567}},
                                          {"$type": "teleport_object",
                                           "id": self.target_objects[1],
                                           "position": {"x": -1.34, "y": 0, "z": 1.894}}])
        self._end_action()
        return ActionStatus.success


if __name__ == "__main__":
    m = Failure(images_directory="D:/transport_challenge_failure_demo", screen_height=1024, screen_width=1024)
    m.init_scene()
    target = m.target_objects[0]
    # Add an overhead camera. Note that `position` is relative to the Magnebot, not absolute worldspace coordinates.
    m.add_camera(position={"x": -1.64, "y": 1.6, "z": -0.166}, look_at=True, follow=True)
    # Collide with a chair.
    m.move_to(target=target)
    # Back away from the chair.
    m.move_by(-0.5, stop_on_collision=False)
    # Run into the chair again.
    m.move_to(target=target)
    # Run into it a little more.
    m.move_by(0.3, stop_on_collision=False)
    # Go around the chair.
    m.move_by(-0.8, stop_on_collision=False)
    m.move_to(target={"x": -0.448, "y": 0, "z": 1.605}, stop_on_collision=False)
    m.turn_to(target=target, stop_on_collision=False)
    # Collide with the table leg or chair.
    m.move_by(0.8, stop_on_collision=False)
    # Try to pick up the object. This sometimes succeeds, and sometimes doesn't (due to Unity physics).
    m.grasp(target=target, arm=Arm.right)
    # Immediately drop the object to simulate a failure.
    if target in m.state.held[Arm.right]:
        m.drop(target=target, arm=Arm.right, wait_for_objects=False)
    m.reset_arm(arm=Arm.right)
    m.move_by(-0.2, arrived_at=0.05, stop_on_collision=False)
    m.pick_up(target=target, arm=Arm.right)
    m.move_by(-0.8, stop_on_collision=False)
    target = m.target_objects[1]
    m.move_to(target=target, stop_on_collision=False)
    m.pick_up(target=target, arm=Arm.left)
    m.move_by(0.5)
    m.end()


controllers/promos/success.py
from transport_challenge import Transport
from magnebot import Arm
from tdw.tdw_utils import TDWUtils

"""
An example of the Magnebot successfully completing a transport task.
"""

if __name__ == '__main__':
    m = Transport(launch_build=False, screen_width=1024, screen_height=1024, fov=90)
    m.init_scene(scene="5a", layout=2, random_seed=9420)
    m.add_camera(position={"x": -2, "y": 0.75, "z": 0}, look_at=True, follow=True)
    # This is the closest container.
    m.move_to(target=m.containers[0])
    m.pick_up(target=m.containers[0], arm=Arm.right)
    # Go to the other room.
    for waypoint in [[-4.91, 0, -0.73], [-4.91, 0, 1.09]]:
        m.move_to(target=TDWUtils.array_to_vector3(waypoint))
    # Pick up some objects.
    for target_object_index in [1, 5, 0]:
        m.move_to(target=m.target_objects[target_object_index], stop_on_collision=False)
        m.pick_up(target=m.target_objects[target_object_index], arm=Arm.left)
        m.put_in()
    # Go to the other room.
    for waypoint in [[-0.49, 0, 4.49], [1.23, 0, 4.49]]:
        m.move_to(target=TDWUtils.array_to_vector3(waypoint))
    # Move to the table.
    for waypoint in [[3.44, 0, 3.09], [4.49, 0, 2.58]]:
        m.move_to(target=TDWUtils.array_to_vector3(waypoint))
    m.pour_out()
    m.reset_arm(arm=Arm.right)
    m.move_by(-1)
    m.end()


controllers/examples/single_room.py
from pathlib import Path
from typing import List
import numpy as np
from tdw.tdw_utils import TDWUtils
from magnebot import ActionStatus, Arm
from transport_challenge import Transport


class SingleRoom(Transport):
    """
    This is an example of how to pick up target objects, put them in a container, and transport them to a goal zone.
    This example does _not_ describe how to implement navigation in the Transport Challenge.
    """

    def get_container(self) -> int:
        """
        This is a VERY naive approach to finding a nearby container. It just returns the nearest one.

        :return: The ID of the nearest container.
        """

        return self._get_nearest(object_ids=self.containers)

    def get_target_object(self) -> int:
        """
        This is a VERY naive approach to finding a nearby target object. It just returns the nearest one.

        :return: The ID of the nearest target object.
        """

        return self._get_nearest(object_ids=self.target_objects)

    def pick_up_container(self) -> None:
        """
        Go to a container in the room and pick it up.
        """

        # Find a container in the room.
        container_id = self.get_container()
        # Go to the container and pick it up.
        self.move_to(target=container_id)
        # Try to pick up the container.
        got_container = False
        while not got_container:
            self.pick_up(target=container_id, arm=Arm.right)
            if container_id in self.state.held[Arm.right]:
                print("Picked up a container.")
                got_container = True
            else:
                print("Failed to pick up a container. Trying again...")
                # Back up and try again.
                # This isn't a very robust approach.
                # In an actual use-case, the actions will vary depending on the position of the object,
                # the position of the Magnebot, etc.
                self.move_by(-0.5)
                self.turn_by(15)
                self.move_to(target=container_id)

    def put_object_in_container(self, object_id: int) -> bool:
        """
        Go to a target object, pick it up, and put it in the container.

        This is a VERY naive function because:

        - It assumes that the Magnebot is holding a container.
        - It doesn't handle navigation; it assumes that there is a clear path between the Magnebot and the object.
        - If the Magnebot fails to put the object in the container, it won't try again.

        :param object_id: The ID of the target object.

        :return: True if the Magnebot put the object in the container.
        """

        # Try to go to the object and pick it up.
        # This is a VERY naive navigation solution. It doesn't check for obstructions.
        status = self.move_to(target=object_id)

        # Back up and try again.
        direction = 1
        num_attempts = 0
        while status != ActionStatus.success and num_attempts < 10:
            print(f"Failed to pick up target object {object_id}. Trying again...")
            self.move_by(-0.5)
            self.turn_by(15 * direction)
            # Alternate the direction.
            direction *= -1
            # Try to go to the object again.
            status = self.move_to(target=object_id)
            num_attempts += 1
        if status != ActionStatus.success:
            print(f"Failed to move to target object {object_id}: {status}")
            return False
        print(f"Moved to target object {object_id}")
        status = self.pick_up(target=object_id, arm=Arm.left)
        if object_id not in self.state.held[Arm.left]:
            print(f"Failed to pick up target object {object_id}: {status}")
            return False
        print(f"Picked up target object {object_id}")
        status = self.put_in()
        if status != ActionStatus.success:
            print(f"Failed to put target object {object_id} in the container: {status}")
            return False
        print(f"Put target object {object_id} in the container.")
        return True

    def _get_nearest(self, object_ids: List[int]) -> int:
        """
        Get the nearest object from a list of object IDs.

        :return: The nearest object in the list.
        """

        nearest = -1
        nearest_distance = np.inf
        for object_id in object_ids:
            distance = np.linalg.norm(self.state.object_transforms[object_id].position -
                                      self.state.magnebot_transform.position)
            if distance < nearest_distance:
                nearest_distance = distance
                nearest = object_id
        return nearest


if __name__ == "__main__":
    # Instantiate the controller.
    # For this example, we set the random seed so that we know where the target objects and containers will be.
    m = SingleRoom(launch_build=False, random_seed=12, auto_save_images=True)
    print(f"Images will be saved to: {Path(m.images_directory).resolve()}")

    # We know that there are target objects and a container in this room because of the random seed in the constructor.
    m.init_scene(scene="5a", layout=2, room=2, goal_room=2)

    m.pick_up_container()

    # Put an object in the container.
    in_container = m.put_object_in_container(object_id=m.get_target_object())
    assert in_container

    # If the container is mostly full, bring it to the goal position and pour it out.
    print("Bringing target object to the goal zone.")
    m.move_to(target=TDWUtils.array_to_vector3(m.goal_position))
    m.pour_out()
    print("Poured out objects.")
    num_in_container = 0

    # Not all of the objects that were in the container are in the goal zone.
    # They might have bounced or rolled away.
    print(f"Objects in the goal zone: {m.get_target_objects_in_goal_zone()}")
    print(f"Action cost: {m.action_cost}")
    m.end()


transport_challenge/promo_controller.py
from typing import Dict, List, Union
from pathlib import Path
from tdw.tdw_utils import TDWUtils
from tdw.output_data import OutputData, Images
from magnebot import ActionStatus
from transport_challenge import Transport


class PromoController(Transport):
    """
    This class is meant to be used for generating promo videos.
    It will automatically save every image per physics simulation step (NOT per-action) from 1-2 cameras (Magnebot and/or overhead).
    This means that this controller will run very slowly.
    """

    def __init__(self, port: int = 1071, screen_width: int = 1024, screen_height: int = 1024,
                 images_directory: str = "images", image_pass_only: bool = False, overhead_camera_only: bool = False,
                 random_seed: int = None):
        super().__init__(port=port, launch_build=False, screen_width=screen_width, screen_height=screen_height,
                         auto_save_images=False, debug=False, images_directory=images_directory, random_seed=random_seed,
                         img_is_png=False, skip_frames=0)

        self.image_directories: Dict[str, Path] = dict()
        self.image_pass_only = image_pass_only
        self.overhead_camera_only = overhead_camera_only
        if not overhead_camera_only:
            self._create_images_directory(avatar_id="a")

        self._image_count = 0

    def add_camera(self, position: Dict[str, float], roll: float = 0, pitch: float = 0, yaw: float = 0,
                   look_at: bool = True, follow: bool = False, camera_id: str = "c") -> ActionStatus:
        """
        See: `Magnebot.add_camera()`.

        Adds some instructions to render images per-frame.
        """

        self._create_images_directory(avatar_id=camera_id)
        status = super().add_camera(position=position, roll=roll, pitch=pitch, yaw=yaw, look_at=look_at, follow=follow,
                                    camera_id=camera_id)
        # Always save images.
        if not self._debug:
            if self.overhead_camera_only:
                self._per_frame_commands.extend([{"$type": "enable_image_sensor",
                                                  "enable": True},
                                                 {"$type": "send_images",
                                                  "ids": ["c"]}])
            else:
                self._per_frame_commands.extend([{"$type": "enable_image_sensor",
                                                  "enable": True},
                                                 {"$type": "send_images"}])
        return status

    def communicate(self, commands: Union[dict, List[dict]]) -> List[bytes]:
        """
        See `Magnebot.communicate()`.

        Images are saved per-frame.
        """

        resp = super().communicate(commands=commands)
        if not self._debug:
            # Save all images.
            got_images = False
            for i in range(len(resp) - 1):
                r_id = OutputData.get_data_type_id(resp[i])
                if r_id == "imag":
                    got_images = True
                    images = Images(resp[i])
                    avatar_id = images.get_avatar_id()
                    if avatar_id in self.image_directories:
                        TDWUtils.save_images(filename=TDWUtils.zero_padding(self._image_count, 8),
                                             output_directory=self.image_directories[avatar_id],
                                             images=images)
            if got_images:
                self._image_count += 1
        return resp

    def _create_images_directory(self, avatar_id: str) -> None:
        """
        :param avatar_id: The ID of the avatar.

        :return: An images directory for the avatar.
        """

        # Build the images directory.
        a_dir = self.images_directory.joinpath(avatar_id)
        if not a_dir.exists():
            a_dir.mkdir(parents=True)
        self.image_directories[avatar_id] = a_dir

    def _get_scene_init_commands(self, magnebot_position: Dict[str, float] = None) -> List[dict]:
        commands = super()._get_scene_init_commands(magnebot_position=magnebot_position)
        # Hide the roof.
        commands.append({"$type": "set_floorplan_roof",
                         "show": False})
        if self.image_pass_only:
            commands.append({"$type": "set_pass_masks",
                             "pass_masks": ["_img"]})
        return commands


controllers/promos/promo.py
from os import chdir
from typing import List
from pathlib import Path
from subprocess import call
from json import loads
import numpy as np
from tdw.tdw_utils import TDWUtils
from magnebot import ActionStatus, Arm
from transport_challenge.promo_controller import PromoController


class Demo(PromoController):
    """
    A demo of a Magnebot using a container to transport target objects to a new room.

    **This is NOT a use-case example.** It should only be used to generate a demo video of the Transport Challenge.

    Key differences:

    - Navigation is pre-calculated (see `PATH`).
    - Only the `img` pass is captured (not `id` or `depth`).
    - The screen is large, there is an overhead camera, and images are saved per-frame instead of per-action. This means that this controller will run *much* slower than a use-case controller.
    - There are some low-level commands to optimize the demo such as teleporting containers and target objects and hiding the roof.
    """

    # This is a pre-calculated path that the Magnebot will use to move between rooms.
    PATH: np.array = np.array([[6.396355, 0, -2.465405],
                               [5.41636, 0, -1.4854207],
                               [4.615, 0, -0.9954208],
                               [3.946356, 0, 0.66],
                               [0.4, 0, 0.66],
                               [0.02635, 0, -1.975]])

    def __init__(self, port: int = 1071, screen_width: int = 1024, screen_height: int = 1024,
                 images_directory: str = "images", image_pass_only: bool = False, overhead_camera_only: bool = False):
        super().__init__(port=port, screen_width=screen_width, screen_height=screen_height,
                         images_directory=images_directory, random_seed=16, image_pass_only=image_pass_only,
                         overhead_camera_only=overhead_camera_only)
        self._to_transport: List[int] = list()

    def init_scene(self, scene: str, layout: int, room: int = None, goal_room: int = None,
                   random_seed: int = None) -> ActionStatus:
        status = super().init_scene(scene=scene, layout=layout, room=room)

        self._to_transport = self.target_objects[:]
        return status

    def transport(self) -> None:
        """
        Transport some objects to the other room.
        """

        for i in range(4):
            # Get the closest object that still needs to be transported.
            self._to_transport = list(sorted(self._to_transport, key=lambda x: np.linalg.norm(
                self.state.object_transforms[x].position - self.state.magnebot_transform.position)))
            object_id = self._to_transport[0]
            # Go to the object and pick it up.
            self.move_to(target=object_id, stop_on_collision=False)
            self.pick_up(target=object_id, arm=Arm.right)
            # Put the object in the container.
            self.put_in()
            # Record this object as done.
            self._to_transport = self._to_transport[1:]

        # Follow the path to the other room.
        path = Demo.PATH[1:]
        for waypoint in path:
            self.move_to(target=TDWUtils.array_to_vector3(waypoint), stop_on_collision=False)
        self.pour_out()

    def go_to_start(self) -> None:
        """
        Navigate back to the start position.
        """

        path = np.flip(Demo.PATH[:-1], axis=0)
        for waypoint in path:
            self.move_to(target=TDWUtils.array_to_vector3(waypoint), stop_on_collision=False)

    def teleport_objects(self) -> None:
        """
        Teleport objects to where they should be for this demo.
        """

        data = loads(Path("init.json").read_text(encoding="utf-8"))
        # Iterate through the containers and the target objects.
        for k, lst in zip(["containers", "target_objects"], [self.containers, self.target_objects]):
            # Set the positions and rotations.
            for object_id, data_object_id in zip(lst, data[k]):
                object_data = data[k][data_object_id]
                self._next_frame_commands.extend([{"$type": "teleport_object",
                                                   "id": object_id,
                                                   "position": object_data["position"]},
                                                  {"$type": "rotate_object_to",
                                                   "id": object_id,
                                                   "rotation": object_data["rotation"]}])
        self._end_action()


if __name__ == "__main__":
    m = Demo(images_directory="D:/transport_challenge_demo", image_pass_only=True, overhead_camera_only=True)
    m.init_scene(scene="2a", layout=1, room=4)
    # Add an overhead camera.
    m.add_camera(position={"x": -3.6, "y": 8, "z": -0.67}, look_at=True, follow=True)

    m.teleport_objects()

    # Pick up the  container.
    m.move_to(target=m.containers[0])
    m.pick_up(target=m.containers[0], arm=Arm.left)

    # Pick up some objects and put them in another room.
    m.transport()
    # Go back to the starting room.
    m.go_to_start()
    m.end()

    # Create a video.
    if m.overhead_camera_only:
        chdir(str(m.image_directories["c"]))
        call(["ffmpeg.exe",
              "-r", "90",
              "-i", "img_%08d.jpg",
              "-vcodec", "libx264",
              "-pix_fmt", "yuv420p",
              str(m.images_directory.joinpath("transport_challenge_demo.mp4").resolve())])
